core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java

```diff
@@ -183,6 +212,95 @@
+     * the specified {@link CircuitBreakerRule}.
+     *
+     * <p>Since {@link CircuitBreaker} is a unit of failure detection, don't reuse the same instance for
+     * unrelated services.
+     *
+     * @param factory a function that takes a method+path and creates a new {@link CircuitBreaker}.
+     */
+    public static Function<? super HttpClient, CircuitBreakerClient>
+    newPerMethodAndPathDecorator(BiFunction<String, String, ? extends CircuitBreaker> factory,
+                                 CircuitBreakerRule rule) {
+        return newDecorator(CircuitBreakerMapping.perMethodAndPath(factory), rule);
+    }
+
+    /**
+     * Creates a new decorator that binds one {@link CircuitBreaker} per request method and path with
+     * the specified {@link CircuitBreakerRuleWithContent}.
+     *
+     * <p>Since {@link CircuitBreaker} is a unit of failure detection, don't reuse the same instance for
+     * unrelated services.
+     *
+     * @param factory a function that takes a method+path and creates a new {@link CircuitBreaker}.
+     */
+    public static Function<? super HttpClient, CircuitBreakerClient>
+    newPerMethodAndPathDecorator(BiFunction<String, String, ? extends CircuitBreaker> factory,
+                                 CircuitBreakerRuleWithContent<HttpResponse> ruleWithContent) {
+        return newDecorator(CircuitBreakerMapping.perMethodAndPath(factory), ruleWithContent);
+    }
+
+    /**
+     * Creates a new decorator that binds one {@link CircuitBreaker} per host, request method and path with
+     * the specified {@link CircuitBreakerRule}.
+     *
+     * <p>Since {@link CircuitBreaker} is a unit of failure detection, don't reuse the same instance for
+     * unrelated services.
+     *
+     * @param factory a function that takes a host+method+path and creates a new {@link CircuitBreaker}.
+     */
+    public static Function<? super HttpClient, CircuitBreakerClient>
+    newPerHostAndMethodAndPathDecorator(CircuitBreakerFactory factory, CircuitBreakerRule rule) {
+        return newDecorator(CircuitBreakerMapping.perHostAndMethodAndPath(factory), rule);
+    }
+
+    /**
+     * Creates a new decorator that binds one {@link CircuitBreaker} per host, request method and path with
+     * the specified {@link CircuitBreakerRuleWithContent}.
+     *
+     * <p>Since {@link CircuitBreaker} is a unit of failure detection, don't reuse the same instance for
+     * unrelated services.
+     *
+     * @param factory a function that takes a host+method+path and creates a new {@link CircuitBreaker}.
+     */
+    public static Function<? super HttpClient, CircuitBreakerClient>
+    newPerHostAndMethodAndPathDecorator(CircuitBreakerFactory factory,
+                                        CircuitBreakerRuleWithContent<HttpResponse> ruleWithContent) {
+        return newDecorator(CircuitBreakerMapping.perHostAndMethodAndPath(factory), ruleWithContent);
+    }
```

Comment:
The factory methods in this class are overly crowded.
I think we can introduce a builder class that builds a `CircuitBreakerMapping` fluently.
For example:
```java
CircuitBreakerMapping mapping = 
    CircuitBreakerMapping.builder()
                         .perMethod()
                         .perPath()
                         ...
                        .build()

CircuitBreakerClient.newDecorator(mapping, rule);
```

Reply:
Done.